ait/core/server/client.py

```diff
@@ -175,3 +245,195 @@ def handle(self, packet, address):
+
+            self.sub = gevent.socket.socket(
+                gevent.socket.AF_INET, gevent.socket.SOCK_STREAM
+            )
+            hostname = (
+                "127.0.0.1" if hostname in ["127.0.0.1", "localhost"] else "0.0.0.0"
+            )
+            super(TCPInputServer, self).__init__(
+                zmq_context,
+                zmq_proxy_xsub_url,
+                zmq_proxy_xpub_url,
+                listener=(hostname, int(port)),
+            )
+        else:
+            raise (ValueError("TCPInputServer:  Invalid Specification"))
+
+    def handle(self, socket, address):
+        self.cur_socket = socket
+        with socket:
+            while True:
+                data = socket.recv(self.buffer)
+                if not data:
+                    break
+                log.debug("{} received message from port {}".format(self, address))
+                self.process(data)
+                gevent.sleep(0)  # pass control back
+
+
+class TCPInputClient(ZMQClient):
+    """
+    This class creates a TCP input client.  Unlike TCPInputServer and UDPInputServer,
+    this class will proactively initiate a connection with an input source and begin
+    receiving data from that source.  This class does not inherit directly from gevent
+    servers and thus implements its own housekeeping functions.  It also implements a
+    start function that spawns a process to stay consistent with the behavior of
+    TCPInputServer and UDPInputServer.
+
+    """
+
+    def __init__(
+        self,
+        zmq_context,
+        zmq_proxy_xsub_url=ait.SERVER_DEFAULT_XSUB_URL,
+        zmq_proxy_xpub_url=ait.SERVER_DEFAULT_XPUB_URL,
+        connection_reattempts=5,
+        buffer=1024,
+        **kwargs,
+    ):
+        self.connection_reattempts = connection_reattempts
+        self.buffer = buffer
+        self.connection_status = -1
+        self.proc = None
+        self.protocol = gevent.socket.SOCK_STREAM
+
+        if "buffer" in kwargs and type(kwargs["buffer"]) == int:
+            self.buffer = kwargs["buffer"]
+
+        if "input" in kwargs:
+            input = kwargs["input"]
+            if not utils.is_valid_address_spec(input):
+                raise (ValueError(f"TCPInputClient: Invalid Specification {input}"))
+            protocol, hostname, port = input.split(":")
+            if protocol.lower() != "tcp":
+                raise (ValueError(f"TCPInputClient: Invalid Specification {input}"))
+            super(TCPInputClient, self).__init__(
+                zmq_context, zmq_proxy_xsub_url, zmq_proxy_xpub_url
+            )
+
+            self.sub = gevent.socket.socket(gevent.socket.AF_INET, self.protocol)
+
+            self.hostname = hostname
+            self.port = int(port)
+            self.address = (hostname, int(port))
+
+        else:
+            raise (ValueError("TCPInputClient: Invalid Specification"))
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        try:
```

Comment:
please consolidate into a single method which is called by exit's and del, and consider Niling the fields after they are closed/killed